### PR TITLE
interrupt_controller: fixed IO APIC broadcasting issue (V2)

### LINKED ISSUE #21156: Interrupts do not work on UP Squared board
**Describe the bug**
`samples/boards/up_squared/gpio_counter` no longer works on UP Squared due to interrupts no longer being delivered.

**To Reproduce**
1. cd samples/boards/up_squared/gpio_counter/
2. mkdir build; cd build
3. cmake -DBOARD=up_squared
4. make
5. Set BIOS configurations described inside `src/main.c`.
6. Boot the Zephyr image according to documentation.
7. Observe the counter would not increment when inputs are pulsed on HAT pin 16.

**Expected behavior**
The counter would increment, and reflected on the 4-bit output of pin35/37/38/40.

**Impact**
- Broken board support.
- Blocking verification of GPIO API changes and app updates.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.10.3
 - Commit: 58e2057aaf9af4d299814f4aece585e287587162

### PR DESCRIPTION
**Pull request by ZideChen0**
This is the updated version for this PR which was closed:
https://github.com/zephyrproject-rtos/zephyr/pull/22848

Currently IO APIC is working in physical destination mode, fixed delivery, and the destination fields in  IOREDTBL is fixed to 0xff, which seems a invalid configuration. By definition, in this setting, IO APIC interrupt can be delivered to one single local APIC whose LAPIC ID is 0xff.

This patchset is intended to fix the IO APIC delivery issue on SMP systems:

- change IO APIC from physical mode to logical destination mode so it's possible to deliver IO APIC interrupts to multiple local APICs.
- HPET timer interrupt needs to run in fixed delivery mode
- other device drivers are good to run in lowest priority mode

Tested:
1) sanitycheck -p qemu_x86
2) sanitycheck -p qemu_x86_64
3) Successfully ran samples/subsys/console/echo (test UART driver) and tests/kernel/queue (test timer interrupts) on these platforms:
- ACRN (x2apic)
- up_squared (both xapic and x2apic, SP and SMP)
- qemu_x86 (xapic)
- qemu_x86_64 (xapic, SP and SMP)

Can't test X2APIC with QEMU due to following warning:
qemu-system-x86_64: warning: TCG doesn't support requested feature: CPUID.01H:ECX.x2apic [bit 21]

fixes #21156

### CODE REVIEW

**Comment by dcpleung** [comment]
This fails to build for `intel_s1000_crb`.

**Comment by ZideChen0** [comment]
> This fails to build for `intel_s1000_crb`.

Fixed.

File: include/dt-bindings/interrupt-controller/intel-ioapic.h [hunk 1]
```diff
@@ -10,9 +10,22 @@
 #define	IRQ_TYPE_EDGE			0x00000000
 #define	IRQ_TYPE_LOW			0x00002000
 #define	IRQ_TYPE_HIGH			0x00000000
-#define IRQ_TYPE_EDGE_RISING		(IRQ_TYPE_EDGE | IRQ_TYPE_HIGH)
-#define IRQ_TYPE_EDGE_FALLING		(IRQ_TYPE_EDGE | IRQ_TYPE_LOW)
-#define IRQ_TYPE_LEVEL_HIGH		(IRQ_TYPE_LEVEL | IRQ_TYPE_HIGH)
-#define IRQ_TYPE_LEVEL_LOW		(IRQ_TYPE_LEVEL | IRQ_TYPE_LOW)
+#define	IRQ_DELIVERY_LOWEST		0x00000100
+#define	IRQ_DELIVERY_FIXED		0x00000000
```
Comment (galak) [inline include/dt-bindings/interrupt-controller/intel-ioapic.h:14]: Why do you need to encode this explicitly?

**Review by galak — DISMISSED** [review]
We should avoid having the DTS interrupt encoding be zephyr specific.  The concept of 'Fixed' and 'Lowest Level' seem to be encode how zephyr encodes irqs.

File: include/dt-bindings/interrupt-controller/intel-ioapic.h (hunk 1 above)
Comment (ZideChen0) [reply include/dt-bindings/interrupt-controller/intel-ioapic.h:14]: Take an UART interrupt as example, all the way from the generated code (e.g. uart_ns16650_port_0.h) to the ioapic driver:

irq_connect_dynamic() -> arch_irq_connect_dynamic() ->z_irq_controller_irq_config() ->z_ioapic_irq_set().

The only way to config different interrupt with different delivery mode (fixed vs lowest priority) is through the irq_flags that is generated from the .dts.

Thus, the easiest approach available is to change these macros as I did in this patch. Other alternatives that I can think of:

1) Don't change the name of IRQ_TYPE_xxx macros, but silently change it's content:
| #define IRQ_TYPE_LEVEL_LOW	(IRQ_TYPE_LEVEL | IRQ_TYPE_LOW|IRQ_DELIVERY_LOWEST) |

2) Add one more cell to the interrupts property. But this seems a big change. For example:
| interrupts = <4 IRQ_TYPE_EDGE_RISING IRQ_DELIVERY_LOWEST_PRIO 3>|

Comment (ZideChen0) [reply include/dt-bindings/interrupt-controller/intel-ioapic.h:14]: Take an UART interrupt as example, all the way from the generated code (e.g. uart_ns16650_port_0.h) to the ioapic driver:

irq_connect_dynamic() -> arch_irq_connect_dynamic() ->z_irq_controller_irq_config() ->z_ioapic_irq_set().

The only way to config different interrupt with different delivery mode (fixed vs lowest priority) is through the irq_flags that is generated from the .dts.

Thus, the easiest approach available is to change these macros as I did in this patch. Other alternatives that I can think of:

Don't change the name of IRQ_TYPE_xxx macros, but silently change it's content:
 #define IRQ_TYPE_LEVEL_LOW (IRQ_TYPE_LEVEL | IRQ_TYPE_LOW|IRQ_DELIVERY_LOWEST) 

Add one more cell to the interrupts property. But this seems a big change. For example:
 interrupts = <4 IRQ_TYPE_EDGE_RISING IRQ_DELIVERY_LOWEST_PRIO 3>

**Comment by ZideChen0** [comment]
> We should avoid having the DTS interrupt encoding be zephyr specific. The concept of 'Fixed' and 'Lowest Level' seem to be encode how zephyr encodes irqs.

Within the current IRQ configuration architect, seems we have to encode the IRQ delivery mode in the .dts. Just need to find a best way to do that. Add a new cell to the interrupts property?

**Comment by galak** [comment]
> Within the current IRQ configuration architect, seems we have to encode the IRQ delivery mode in the .dts. Just need to find a best way to do that. Add a new cell to the interrupts property?

what is the delivery mode mean?

**Comment by ZideChen0** [comment]
> > Within the current IRQ configuration architect, seems we have to encode the IRQ delivery mode in the .dts. Just need to find a best way to do that. Add a new cell to the interrupts property?
> 
> what is the delivery mode mean?

Page 13 in:
https://pdos.csail.mit.edu/6.828/2014/readings/ia32/ioapic.pdf

Basically it's instructing IO APIC to deliver the external interrupt to all CPUs or just to one single CPU.

**Comment by galak** [comment]
Yeah, an additional cell to encode this would make sense.

**Review by tyrantlizard77 — APPROVED** [review]
Code changes LGTM.

**Comment by galak** [comment]
Would be good to update `dts/bindings/interrupt-controller/intel,ioapic.yaml` to provide a bit more detail about the encoding.  You can look at `dts/bindings/dma/st,stm32-dma.yaml ` for an example of describing encodes in description.

**Comment by nashif** [comment]
enhancement created #25160

**Review by nashif — APPROVED** [review]

**Pull request merged**